README.md
# Hotline
Create connections to an ActionCable server with an ActionCable-like subscriptions API. 

## Dart example

```dart
void main() {
  // Create a connection to an ActionCable server
  final hotline = Hotline('ws://localhost:3000/cable');

  // Next, subscribe to a channel
  // ActionCable.server.broadcast("Chat Room", {hello: "Chat room!"})
  // `simpleChannel` onReceived callback will be called when subscribing
  // to a simple channel name using `stream_from`
  final subscription = hotline.subscriptions.create('Example Channel');

  // If you want to broadcast to a resource-specific channel, like this:
  // ChatRoomChannel.broadcast_to(ChatRoom.find(1), {hello: "Chat Room 1"})
  // then use a parameterised constructor that and pass the params into your `stream_for` method
  final subscription = hotline.subscriptions.create({'channel': 'Chat Room', 'id': 1});

  // Or even a Turbo Stream
  final subscription = hotline.subscriptions.create({
    'channel': 'Turbo::StreamsChannel',
    'signed_stream_name': 'IloybGtPaTh2WVhCd0wwSmhjMnRsZEM4eCI=--635611ff87dda934376806f3dbfcdb6b724fa0cbcca9b32684c537ad8eb9390f'
  });

  // Finally, you can listen for the various changes on the connection and the channel itself...
  hotline.status.listen((state) => print('Hotline Connection State: $state'));
  subscription.status.listen((state) => print('Subscription State: $state'));
  subscription.stream.listen((data)  => print('Received On Stream: $data'));
}
```

## Dart example using callbacks rather than streams

```dart
void main() {
  // create a connection with the necessary callbacks 
  final onConnect = (Hotline connection) => print('Hotline Connection State: Connected');
  final onDisconnect = () => print('Hotline Connection State: Disconnected');
  Hotline hotline = Hotline('ws://localhost:3000/cable',
      onConnect: onConnect,
      onDisconnect: onDisconnect
  );

  // Create a subscription to one of your ActionCable channels
  final confirmed = (HotlineSubscription subscription) => print('Subscription State: Confirmed');
  final received  = (Map data) => print('Received On Stream: $data');
  final rejected  = () => print('Subscription State: Rejected');

  // ActionCable.server.broadcast("Chat Room", {hello: "Chat room!"})
  // `simpleChannel` onReceived callback will be called when subscribing
  // to a simple channel name using `stream_from`
  HotlineSubscription simpleChannel = hotline.subscriptions.create(
      'Chat Room',
      onConfirmed: confirmed,
      onReceived: received,
      onRejected: rejected
  );

  // If you want to broadcast to a resource-specific channel, like this:
  // ChatRoomChannel.broadcast_to(ChatRoom.find(1), {hello: "Chat Room 1"})
  // then use a parameterised constructor and pass the params into your `stream_for` method 
  HotlineSubscription resourceChannel = hotline.subscriptions.create(
      {'channel': 'Chat Room', 'id': 1},
      onConfirmed: confirmed,
      onReceived: received,
      onRejected: rejected
  );
}
```

## Flutter

### It's recommended that you use the stream-based interface, with StreamBuilder widgets. 

```dart
// Create the container widget for instantiating our connection
class HotlineConnectionView extends StatefulWidget {
  final hotline = Hotline('ws://localhost:3000/cable');

  HotlineConnectionView({Key? key}) : super(key: key);

  @override
  State<HotlineConnectionView> createState() => _HotlineConnectionViewState();
}

class _HotlineConnectionViewState extends State<HotlineConnectionView> {
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return StreamBuilder(
      stream: this.widget.hotline.status,
      builder: (BuildContext context, AsyncSnapshot<HotlineSocketConnectionType> snapshot) {
        if(snapshot.data == HotlineSocketConnectionType.connecting) {
          // whilst we're waiting for HotlineSocketConnectionType.connected, show a spinner
          return CircularProgressIndicator();
        }else if(HotlineSocketConnectionType.connected) {
          // if we're here, we can render the HotlineSubscriptionView for the channel 
          // we want to subscribe to, and pass the hotline instance along with it
          return HotlineSubscriptionView({'channel': 'Chat Room', 'id': 1}, this.widget.hotline);
        }else {
          return Text('Not connected');
        }
      }
    );
  }
}

// A widget for creating a HotlineSubscription to the given channel
class HotlineSubscriptionView extends StatefulWidget {
  late final HotlineSubscription subscription;
  HotlineSubscriptionView(Map<String, String> channel, Hotline connection) : subscription = connection.subscriptions.create(channel);

  @override
  _HotlineSubscriptionViewState createState() => _HotlineSubscriptionViewState();
}

class _HotlineSubscriptionViewState extends State<HotlineSubscriptionView> {
  @override
  Widget build(BuildContext context) {
    return Container(
      child: Column(
        children: [
          // StreamBuilder to show the current connection status to the channel
          StreamBuilder(
            stream: this.widget.subscription.status,
            builder: (BuildContext context, AsyncSnapshot<HotlineSubscriptionRequestState> snapshot) {
              return Text('Subscription Status: ${snapshot.data}');
            },
          ),
          // Another StreamBuilder, this time to render data as it comes down the wire
          StreamBuilder(
            stream: this.widget.subscription.stream,
            builder: (BuildContext context, AsyncSnapshot<Map> snapshot) {
              return Text('Data: ${snapshot.data}');
            },
          ),
        ],
      )
    );
  }
}
```

## Thanks
Hotline borrows some implementation from [ActionCable in Dart] by [https://github.com/namiwang]

[ActionCable in Dart]: https://pub.dev/packages/action_cable
[https://github.com/namiwang]: https://github.com/namiwang


CHANGELOG.md
## 1.0.0

- Initial version

## 1.0.1

- Fix pedantic lints

## 1.0.2

- Fix more dart syntax
- Added docs

## 1.0.3

- dartfmt formatting fixes

## 1.0.4

- Pass headers and params onto IOWebSocketChannel constructor
- Updated example docs

## 1.1.0

- Added ability to listen to Hotline and HotlineSubscription instances as streams
- Made all callbacks optional parameters
- Updated docs


test/hotline_test.dart
import 'package:hotline/src/hotline_connection_state.dart';
import 'package:hotline/src/hotline_subscription.dart';
import 'package:test/test.dart';
import 'package:hotline/hotline.dart';

import 'test_helpers.dart';

void main() {
  var receivedCallback = 0;

  final onConfirmedCallback = (HotlineSubscription subscription) {};
  final onReceivedCallback = (Map data) {
    receivedCallback += 1;
  };

  group('Connections', () {
    late MockHotline connection;

    setUp(() {
      connection = MockHotline(onConnect: () => {}, onDisconnect: () => {});
    });

    tearDown(() {
      connection.disconnect();
    });

    group('Connection Callbacks', () {
      test('It receives a confirmation callback', () {
        expect(connection.connectionState.state,
            HotlineSocketConnectionType.connecting);

        connection.dispatch({
          'type': 'welcome',
          'identifier': {'channel': 'Some Channel'}
        });

        expect(connection.connectionState.state,
            HotlineSocketConnectionType.connected);
      });

      test('It can be disconnected remotely', () async {
        expect(connection.connectionState.state,
            HotlineSocketConnectionType.connecting);

        connection.dispatch({
          'type': 'welcome',
          'identifier': {'channel': 'Some Channel'}
        });

        connection.dispatch({'type': 'disconnect'});

        expect(connection.connectionState.state,
            HotlineSocketConnectionType.disconnected);
      });
    });
  });

  group('Subscriptions', () {
    late MockHotline connection;

    setUp(() {
      connection = MockHotline(onConnect: () => {}, onDisconnect: () => {});
    });

    group('Subscription Callbacks', () {
      test('It creates a subscription', () {
        final _ = connection.subscriptions.create('Some Channel',
            onReceived: onReceivedCallback, onConfirmed: onConfirmedCallback);

        expect(connection.subscriptions.subscriptions.length, 1);
      });

      test('It confirms a subscription', () {
        final subscription = connection.subscriptions.create('Some Channel',
            onReceived: onReceivedCallback, onConfirmed: onConfirmedCallback);

        expect(subscription.state == HotlineSubscriptionRequestState.granted,
            false);

        connection.dispatch({
          'type': 'confirm_subscription',
          'identifier': {'channel': 'Some Channel'}
        });
        expect(subscription.state, HotlineSubscriptionRequestState.granted);
      });

      test('It receives data', () {
        receivedCallback = 0;
        final _ = connection.subscriptions.create('Some Channel',
            onConfirmed: onConfirmedCallback, onReceived: onReceivedCallback);

        connection.dispatch({
          'message': 'confirm_subscription',
          'identifier': {'channel': 'Some Channel'}
        });

        expect(receivedCallback, 1);
      });

      test('It allows for multiple subscriptions to the same channel', () {
        final _ = connection.subscriptions.create('Some Channel',
            onReceived: onReceivedCallback, onConfirmed: onConfirmedCallback);
        final __ = connection.subscriptions.create('Some Channel',
            onReceived: onReceivedCallback, onConfirmed: onConfirmedCallback);

        connection.dispatch({
          'type': 'confirm_subscription',
          'identifier': {'channel': 'Some Channel'}
        });
        connection.dispatch({
          'type': 'confirm_subscription',
          'identifier': {'channel': 'Some Channel'}
        });

        expect(connection.subscriptions.subscriptions.length, 2);
      });

      test('It doesn\t dispatch messages from suspended subscriptions', () {
        receivedCallback = 0;

        var subscription = connection.subscriptions.create('Test Channel',
            onReceived: (Map data) {
          receivedCallback += 1;
          print('received data: $data - $receivedCallback');
        }, onConfirmed: onConfirmedCallback);

        expect(subscription.state == HotlineSubscriptionRequestState.granted,
            false);

        connection.dispatch({
          'type': 'confirm_subscription',
          'identifier': {'channel': 'Test Channel'}
        });

        expect(subscription.state, HotlineSubscriptionRequestState.granted);

        subscription.suspend();

        connection.dispatch({
          'identifier': {'channel': 'Test Channel'},
          'something': 'hello'
        });

        expect(receivedCallback, 0);
      });
    });

    group('Multiple Subscriptions', () {
      test('It allows duplicate subscriptions to receive data', () {
        receivedCallback = 0;

        final _ = connection.subscriptions.create('Some Channel',
            onConfirmed: onConfirmedCallback, onReceived: onReceivedCallback);
        final __ = connection.subscriptions.create('Some Channel',
            onConfirmed: onConfirmedCallback, onReceived: onReceivedCallback);

        connection.dispatch({
          'message': 'confirm_subscription',
          'identifier': {'channel': 'Some Channel'}
        });

        expect(receivedCallback, connection.subscriptions.subscriptions.length);
      });
    });

    group('Subscription Manager', () {
      test(
          'It allows subscriptions to be unsubscribed by the subscription manager',
          () {
        final subscription = connection.subscriptions.create('Some Channel',
            onConfirmed: onConfirmedCallback, onReceived: onReceivedCallback);
        connection.dispatch({
          'message': 'confirm_subscription',
          'identifier': {'channel': 'Some Channel'}
        });

        connection.subscriptions.unsubscribe(subscription);
        expect(connection.subscriptions.subscriptions.length, 0);
      });

      test('It allows subscriptions to unsubscribe themselves', () {
        final subscription = connection.subscriptions.create('Some Channel',
            onConfirmed: onConfirmedCallback, onReceived: onReceivedCallback);
        connection.dispatch({
          'message': 'confirm_subscription',
          'identifier': {'channel': 'Some Channel'}
        });

        subscription.unsubscribe();
        expect(connection.subscriptions.subscriptions.length, 0);
      });
    });

    group('Subscription Rejections', () {
      test('It handles rejected subscriptions', () async {
        final subscription = connection.subscriptions.create('Some Channel',
            onConfirmed: onConfirmedCallback, onReceived: onReceivedCallback);
        connection.dispatch({
          'message': 'reject_subscription',
          'identifier': {'channel': 'Some Channel'}
        });

        expect(connection.subscriptions.subscriptions.length, 1);

        expect(connection.subscriptions.subscriptions.where((subscription) {
          return subscription.state ==
              HotlineSubscriptionRequestState.subscribing;
        }), [subscription]);

        // wait for subscription confirmation callback to fail and check that it is removed from the pool
        expect(
            await Future.delayed(Duration(milliseconds: 550), () {
              return connection.subscriptions.subscriptions.length;
            }),
            0);
      });

      test('It received an onRejected callback', () async {
        var callbackCount = 0;

        final _ = connection.subscriptions.create('Some Channel',
            onConfirmed: onConfirmedCallback,
            onReceived: onReceivedCallback,
            onRejected: () => callbackCount += 1);

        connection.dispatch({
          'message': 'reject_subscription',
          'identifier': {'channel': 'Some Channel'}
        });

        expect(
            await Future.delayed(
                Duration(milliseconds: 550), () => callbackCount),
            1);
      });
    });
  });
}


test/test_helpers.dart
import 'dart:convert';

import 'package:hotline/src/hotline_connection_state.dart';
import 'package:mockito/mockito.dart';

import 'package:hotline/hotline.dart';
import 'package:hotline/src/hotline_subscription_manager.dart';
import 'package:web_socket_channel/io.dart';

class MockHotline extends Mock implements Hotline {
  @override
  late final HotlineSocketConnectionState connectionState;

  late HotlineSubscriptionManager subscriptions;
  @override
  IOWebSocketChannel socketChannel = IOWebSocketChannel.connect('ws://localhost');

  Function? onConnect;
  Function? onDisconnect;

  MockHotline({required this.onConnect, required this.onDisconnect}) {
    subscriptions = HotlineSubscriptionManager(this);
    connectionState = HotlineSocketConnectionState(
        onConnect: onConnect, onDisconnect: onDisconnect);
    connectionState.stateType = HotlineSocketConnectionType.connecting;
  }

  void dispatch(Map payload) {
    if (payload['type'] == null) {
      subscriptions
          .getAllSubscriptions(jsonEncode(payload['identifier']))
          .forEach((subscription) {
        subscription.handleResponse(payload);
      });
    } else {
      switch (payload['type']) {
        case "welcome":
          this._onConnect();
          break;
        case "disconnect":
          this._onDisconnect();
          break;
        case "confirm_subscription":
          subscriptions.confirmSubscription(jsonEncode(payload["identifier"]));
          break;
        case "reject_subscription":
          final callback = this.onConnectionRefused;
          if (callback != null) {
            callback();
          }
          break;
      }
    }
  }

  void _onConnect() {
    connectionState.stateType = HotlineSocketConnectionType.connected;
    final callback = onConnect;
    if(callback != null) callback();
  }

  void _onDisconnect() {
    connectionState.stateType = HotlineSocketConnectionType.disconnected;
    final callback = onDisconnect;
    if(callback != null) callback();
  }

  void disconnect() {
    connectionState.stateType = HotlineSocketConnectionType.disconnected;
  }
}


lib/src/hotline_base.dart
library hotline;

import 'dart:async';
import 'dart:convert';
import 'package:web_socket_channel/io.dart';

import 'hotline_subscription_manager.dart';
import 'hotline_connection_state.dart';

// Hotline - an ActionCable-like API to subscribe to ActionCable channels
class Hotline {
  /// WebSocket url
  String url;

  /// health-check manager and subscription interface
  late HotlineSocketConnectionState connectionState;
  late StreamSubscription stream;
  late final HotlineSubscriptionManager subscriptions;

  /// optional headers
  Iterable<String>? protocols;
  Map<String, String>? headers;

  /// WebSocket connection callbacks
  Function? onConnect;
  Function? onDisconnect;
  Duration? pingInterval;
  Function? onConnectionRefused;

  /// The wrapped socket channel
  late IOWebSocketChannel socketChannel;

  Hotline(this.url, {
        this.onConnect,
        this.onDisconnect,
        this.headers,
        this.protocols,
        this.pingInterval,
        this.onConnectionRefused
      })
  {
    connectionState = HotlineSocketConnectionState(onConnect: _onConnect, onDisconnect: _onDisconnect);
    connectionState.stateType = HotlineSocketConnectionType.connecting;

    _startWebSocketListener();

    /// Initialise the subscription manager for this connection
    subscriptions = HotlineSubscriptionManager(this);
  }

  /// A Stream for listeners of the connection state
  Stream<HotlineSocketConnectionType> get status => connectionState.status;

  void reconnect() {
    IOWebSocketChannel.connect(
        url,
        protocols: protocols,
        headers: headers,
        pingInterval: pingInterval
    );
    connectionState = HotlineSocketConnectionState(
        onConnect: _onConnect,
        onDisconnect: _onDisconnect
    );

    connectionState.stateType = HotlineSocketConnectionType.connecting;

    _startWebSocketListener();

    subscriptions.unsuspendAll();
  }

  void _startWebSocketListener() {
    socketChannel = IOWebSocketChannel.connect(
        url,
        protocols: protocols,
        headers: headers,
        pingInterval: pingInterval
    );

    stream = socketChannel.stream.listen((data) {
      final payload = jsonDecode(data);

      if (payload['type'] != null) {
        // if the payload has a type attribute we can assume it's a protocol message
        _dispatchProtocolMessage(payload);
      } else {
        // handle the broadcast....
        _dispatchDataMessage(payload);
      }
    }, onError: (_) {
      _onDisconnect();
    });
  }

  void _onConnect() {
    final callback = onConnect;

    if(callback != null) {
      callback(this);
    }
  }

  void _onDisconnect() {
    final callback = onDisconnect;

    connectionState.stateType = HotlineSocketConnectionType.disconnected;

    if (connectionState.isConnected) {
      socketChannel.sink.close();
    }

    stream.cancel();

    if(callback != null) {
      callback();
    }
  }

  // connection response dispatcher.
  void _dispatchProtocolMessage(Map payload) {
    switch (payload['type']) {
      case 'welcome':
        connectionState.stateType = HotlineSocketConnectionType.connected;
        break;
      case 'disconnect':
        connectionState.stateType = HotlineSocketConnectionType.disconnected;
        break;
      case 'confirm_subscription':
        subscriptions.confirmSubscription(payload['identifier']);
        break;
      case 'reject_subscription':
        final callback = onConnectionRefused;

        if (callback != null) {
          callback();
        }

        break;
      case 'ping':
        connectionState.lastPing = DateTime.fromMillisecondsSinceEpoch(payload['message'] * 1000);
        break;
    }
  }

  void _dispatchDataMessage(Map payload) {
    // get our subscription from the subscription manager and
    // pass the payload on to it...
    final allSubscriptions = subscriptions.getAllSubscriptions(payload['identifier']);
    allSubscriptions.forEach((subscription) => subscription.handleResponse(payload));
  }

  /// Temporarily stop dispatching to any subscriptions
  void suspend() {
    subscriptions.suspendAll();
  }

  /// Disconnect from the server and close any subscriptions
  void disconnect() {
    connectionState.stateType = HotlineSocketConnectionType.disconnected;
    socketChannel.sink.close();
  }
}


lib/src/hotline_subscription.dart
import 'dart:async';
import 'dart:convert';
import 'hotline_subscription_manager.dart';

/// The valid states for a subscription
enum HotlineSubscriptionRequestState {
  initial,
  subscribing,
  unsubscribed,
  granted,
  rejected,
  suspended
}

/// A subscription instance for handling channel-specific messages
class HotlineSubscription {
  String identifier;
  HotlineSubscriptionManager subscriptionManager;
  HotlineSubscriptionRequestState state = HotlineSubscriptionRequestState.initial;

  final StreamController<Map> _controller = StreamController<Map>();
  final StreamController<HotlineSubscriptionRequestState> _state_controller = StreamController<HotlineSubscriptionRequestState>();

  /// set the state of this subscription
  set stateType(HotlineSubscriptionRequestState type) => state = type;

  /// this dispatch the data we receive onto the stream listeners...
  Stream<Map> get stream async* {
    yield* _controller.stream;
  }

  /// anything listening for updates to the subscription request...
  Stream<HotlineSubscriptionRequestState> get status async* {
    yield* _state_controller.stream;
  }

  /// optional subscription callbacks
  Function? onReceived;
  Function? onConfirmed;
  Function? onUnsubscribed;
  Function? onRejected;

  final _SUBSCRIPTION_REQUEST_TIMEOUT = 500;

  /// when cancelled by the subscription manager no additional work is required to remove this from the pool
  var _cancelledBySubscriptionManager = false;
  set cancelledBySubscriptionManager(bool cancelled) => _cancelledBySubscriptionManager = cancelled;

  HotlineSubscription(this.identifier, this.subscriptionManager, {required this.onReceived, required this.onConfirmed, this.onUnsubscribed, this.onRejected}) {
    final subscriptionParameters = jsonEncode({'identifier': identifier, 'command': 'subscribe'});

    /// dispatch the subscription
    _sendSubscriptionRequest(subscriptionParameters);
    stateType = HotlineSubscriptionRequestState.subscribing;

    /// if the subscription hasn't been accepted with n milliseconds, assume it's been rejected or the server has gone away
    Future.delayed(Duration(milliseconds: _SUBSCRIPTION_REQUEST_TIMEOUT), () {
      if (state != HotlineSubscriptionRequestState.granted) rejected();
    });
  }

  /// dispatch a new subscription request
  void _sendSubscriptionRequest(String payload) =>
      subscriptionManager.connection.socketChannel.sink.add(payload);

  /// called by the Hotline connection dispatcher
  void confirmed() {
    state = HotlineSubscriptionRequestState.granted;

    final callback = onConfirmed;

    if(callback != null) {
      callback(this);
    }

    _state_controller.add(state);
  }

  /// stop receiving messages for this channel
  void unsubscribe() {
    final callback = onUnsubscribed;
    final unsubscribeParameters = jsonEncode({
      'identifier': identifier,
      'command': 'unsubscribe'
    });

    /// ask the subscriptionManager to unsubscribe so that everything is cleaned up properly
    if (!_cancelledBySubscriptionManager) {
      subscriptionManager.unsubscribe(this);

      subscriptionManager.connection.socketChannel.sink.add(unsubscribeParameters);
      state = HotlineSubscriptionRequestState.unsubscribed;

      _state_controller.add(state);
    } else if (callback != null) {
      callback();
    }
  }

  /// the subscription request was not granted
  void rejected() {
    final callback = onRejected;
    state = HotlineSubscriptionRequestState.rejected;
    _state_controller.add(state);

    if (callback != null) {
      callback();
    }

    unsubscribe();
  }

  /// determine how to handle this response and call the relevant handler...
  void handleResponse(Map payload) {
    if (state == HotlineSubscriptionRequestState.suspended) return;

    final callback = onReceived;

    if(callback != null) {
      callback(payload);
    }

    _controller.add(payload);
  }

  /// send a message to a WebSocket channel on the server to its corresponding channel [action]
  void perform(String action, [Map<String, dynamic>? params]) {
    if (state == HotlineSubscriptionRequestState.suspended) return;

    final actionParams = {'action': action, 'id': 1};

    subscriptionManager.connection.socketChannel.sink.add(jsonEncode({
      'identifier': identifier,
      'command': 'message',
      'data': jsonEncode(actionParams),
    }));
  }

  /// temporarily stop receiving data on this subscription instance
  void suspend() {
    stateType = HotlineSubscriptionRequestState.suspended;
  }
}


lib/src/hotline_connection_state.dart
import 'dart:async';

/// The possible states for a HotlineConnectionState
enum HotlineSocketConnectionType {
  initial,
  connecting,
  connected,
  disconnected
}

/// State manager for a Hotline connection
class HotlineSocketConnectionState {
  Function? onConnect;
  Function? onDisconnect;

  // Vars for periodically check the status of the connection
  Timer? _timer;
  DateTime? _lastPing;
  HotlineSocketConnectionType state = HotlineSocketConnectionType.initial;

  // Stream controller for sending connection state events
  final StreamController<HotlineSocketConnectionType> _controller = StreamController<HotlineSocketConnectionType>();

  HotlineSocketConnectionState({required this.onConnect, required this.onDisconnect});

  /// setter to change our state and trigger additional effects such as our
  /// onConnect callback, or cancelling the health-check timer
  set stateType(HotlineSocketConnectionType type) {
    state = type;

    switch (state) {
      case HotlineSocketConnectionType.initial:
        _controller.add(HotlineSocketConnectionType.initial);
        break;
      case HotlineSocketConnectionType.connecting:
        _controller.add(HotlineSocketConnectionType.connecting);
        break;
      case HotlineSocketConnectionType.connected:
        _controller.add(HotlineSocketConnectionType.connected);
        _startTimer();
        _handleConnectedEvent();
        break;
      case HotlineSocketConnectionType.disconnected:
        _controller.add(HotlineSocketConnectionType.disconnected);
        _stopTimer();
        break;
    }
  }

  /// check whether the connection is still connected
  bool get isConnected => state == HotlineSocketConnectionType.connected;

  /// Set the latest ping time from the socket
  set lastPing(DateTime ping) => _lastPing = ping;

  /// allow for listeners on the connection state
  Stream<HotlineSocketConnectionType> get status async* {
    yield HotlineSocketConnectionType.initial;
    yield* _controller.stream;
  }

  /// periodically check the ping times to ensure we're still connected
  void _startTimer() {
    _timer = Timer.periodic(const Duration(seconds: 3), _connectionHealthCheck);
  }
  
  void _stopTimer() {
    _timer?.cancel();
  }

  void _handleConnectedEvent() {
    final callback = onConnect;
    if(callback != null) {
      callback();
    }
  }

  void _connectionHealthCheck(_) {
    if (_lastPing == null) return;

    if (DateTime.now().difference(_lastPing!) > Duration(seconds: 5)) {
      final callback = onDisconnect;

      if(callback != null) {
        callback();
      }
    }
  }
}


lib/src/hotline_subscription_manager.dart
import 'dart:convert';

import 'package:hotline/hotline.dart';

/// Manages the subscriptions for a connection
class HotlineSubscriptionManager {
  final List<HotlineSubscription> _subscriptions;
  Hotline connection;

  HotlineSubscriptionManager(this.connection) : _subscriptions = [];

  /// Create a new subscription to a channel, using a dynamic [channel] type to allow for simple String subscriptions or complex channels with parameters.
  ///
  /// If using stream_for, with resource specific channels use `{'channel': 'ChannelName', 'param': 1}`
  /// alternatively if you're broadcasting via ActionCable.server.broadcast('ChannelName', {payload: ...}) and `stream_from 'ChannelName'` [channel] should be a string
  HotlineSubscription create(
      dynamic channel,
      {
        Function? onReceived,
        Function? onConfirmed,
        Function? onUnsubscribed,
        Function? onRejected
      })
  {
    final identifier = _getChannelIdentifier(channel);
    final subscription = HotlineSubscription(
        identifier,
        this,
        onReceived: onReceived,
        onConfirmed: onConfirmed,
        onRejected: onRejected,
        onUnsubscribed: onUnsubscribed
    );

    _subscriptions.add(subscription);

    return subscription;
  }

  /// Unsubscribe from the messaging for a given channel
  void unsubscribe(HotlineSubscription subscription) {
    subscription.cancelledBySubscriptionManager = true;
    subscription.unsubscribe();
    _subscriptions.remove(subscription);
  }

  List<HotlineSubscription> get subscriptions => _subscriptions;

  /// allow for `hotline.consumer.subscriptions[0]`
  HotlineSubscription operator [](int i) => _subscriptions[i];

  /// turn the channel identifier into an encoded string that ActionCable is expecting
  String _getChannelIdentifier(dynamic identifier) {
    if (identifier is Map<String, dynamic>) {
      return jsonEncode(identifier);
    } else {
      return jsonEncode({'channel': identifier});
    }
  }

  /// Find a subscription from the collection by looking up its identifier
  HotlineSubscription? getSubscription(String identifier) {
    try {
      return subscriptions.firstWhere((subscription) => subscription.identifier == identifier);
    } catch (e) {
      return null;
    }
  }

  /// return a list of all subscriptions that match a given [identifier]
  ///
  /// allows for subscribing to a channel multiple times
  List<HotlineSubscription> getAllSubscriptions(String identifier) {
    return subscriptions
        .where((subscription) => subscription.identifier == identifier)
        .toList();
  }

  /// called when our `confirmed` callback is triggered by a 'confirm_subscription' event
  void confirmSubscription(String identifier) {
    subscriptions.where((subscription) => subscription.identifier == identifier)
        .forEach((subscription) => subscription.confirmed());
  }

  void suspendAll() {
    subscriptions.forEach((subscription) {
      subscription.state = HotlineSubscriptionRequestState.suspended;
    });
  }

  void unsuspendAll() {
    subscriptions.forEach((subscription) {
      subscription.state = HotlineSubscriptionRequestState.granted;
    });
  }
}


example/hotline_example.dart
import 'package:hotline/hotline.dart';

void main() async {
  final hotline = Hotline('ws://localhost:3000/cable');
  final subscription = hotline.subscriptions.create({
    'channel': 'Turbo::StreamsChannel',
    'signed_stream_name': 'IloybGtPaTh2WVhCd0wwSmhjMnRsZEM4eCI=--635611ff87dda934376806f3dbfcdb6b724fa0cbcca9b32684c537ad8eb9390f'
  });

  /// listen for changes on the connection state...
  hotline.status.listen((state) => print('Hotline Connection State: $state'));
  /// listen for changes on the subscription state...
  subscription.status.listen((state) => print('\tSubscription State: $state'));
  /// listen for data being dispatched to the subscription...
  subscription.stream.listen((data)  => print('\t\tReceived On Stream: \n\t$data'));

  /// ActionCable.server.broadcast("Example Channel", {hello: "Chat room!"})
  /// `simpleChannel` onReceived callback will be called when subscribing
  /// to a simple channel name using `stream_from`
  /// final subscription = hotline.subscriptions.create('Example Channel');

  /// If you want to broadcast to a resource-specific channel, like this:
  /// ChatRoomChannel.broadcast_to(ChatRoom.find(1), {hello: "Chat Room 1"})
  /// then use a parameterised constructor that and pass the params into your `stream_for` method

  /*
  Hotline with callbacks
   */
  /// create a connection with the necessary callbacks
  final onConnect = (Hotline connection) => print('Connected!!!');
  final onDisconnect = () => print('Disconnected');

  final hotlineWithCallbacks = Hotline('ws://localhost:3000/cable', onConnect: onConnect, onDisconnect: onDisconnect);

  /// Create a subscription to one of your ActionCable channels
  final confirmed = (HotlineSubscription subscription) => print('Subscription Confirmed ${subscription}');
  final received = (Map data) => print('Received: $data');
  final rejected = () => print('Subscription Rejected');
  final _ = hotlineWithCallbacks.subscriptions.create({
    'channel': 'Turbo::StreamsChannel',
    'signed_stream_name': 'IloybGtPaTh2WVhCd0wwSmhjMnRsZEM4eCI=--635611ff87dda934376806f3dbfcdb6b724fa0cbcca9b32684c537ad8eb9390f'
  },
      onConfirmed: confirmed,
      onReceived: received,
      onRejected: rejected
  );

  /// Disconnect from the server
  // Future.delayed(Duration(milliseconds: 250), () {
  //   hotline.disconnect();
  //   print(hotline.connectionState.state);
  // });
  //
  // /// try to reconnect an existing connection
  // Future.delayed(Duration(milliseconds: 500), () {
  //   hotline.reconnect();
  // });
}
